### PR TITLE
add permutecols and extend permutecols!

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/1942

@nilshg - can you please test this PR for bugs and if it follows what you would expect?

@nalimilan - as usual apart from general comments a review of documentation would be appreciated.

Note that earlier `permutecols` made no sense as it was the same as `df[:, perm]` but now it makes sense, as we accept subsets of columns.

### CODE REVIEW

**Comment by bkamins** [comment]
That is exactly the point of difference of `permutecols!` and `permutecols` - this example is intended to show the difference between them.

Call `permutecols!(df, [:b, :a])` takes the data frame with columns `[:c, :a, :b]`, because earlier `permutecols!` changed the column order in `df` data frame.

And `permutecols(df, [:b, :a])` in the second takes the data frame with columns `[:a, :b, :c]`, as earlier `permutecols` did not alter the column order of `df`.

**Comment by bkamins** [comment]
Now I have had a look at Stata documentation of `order`. We might add a kwarg `location` that takes a default value of `nothing` which does what I currently implemented, then it can take following values:
* `:first` - to put them in front of the data frame
* `:last` - to put them at the end of the data frame
* `:before=>column_name`
* `:after=>column_name`

where `column_name` is a `Symbol` giving the name of the column. @nalimilan - what do you think about it? If we added all this we would get "more or less" all that is potentially needed. The only question is if this should be still called `permutecols!` 😄.

**Comment by bkamins** [comment]
actually it is even simpler to implement `location` as a last positional argument (so I would opt for this). Any preference here.

**Comment by bkamins** [comment]
OK. I propose to rename the functions to `reordercols!` and `reordercols`.
I have implemented all the functionality I have proposed.

I would appreciate if you did some testing if you like it and commented. If we agree that this is what we want then I will write tests of the functionality and update docstrings.

**Comment by nilshg** [comment]
What's the syntax in Pkg to install a specific fork?

**Comment by bkamins** [comment]
use `PackageSpec` to point to url and branch. But as this change is very simple it is easier to `dev` normal package and just copy the definitions into the appropriate file (this is a bit hacky but in this case it is simplest as the code-block to copy is a simple copy-paste).

File: src/dataframe/dataframe.jl [hunk 1]
```diff
@@ -1358,11 +1358,13 @@ function Base.push!(df::DataFrame, row::Any)
...
 
 """
-    permutecols!(df::DataFrame, p::AbstractVector)
+    reordercols!(df::DataFrame, cols::AbstractVector)
 
-Permute the columns of `df` in-place, according to permutation `p`. Elements of `p` may be
-either column indices (`Int`) or names (`Symbol`), but cannot be a combination of both. All
-columns must be listed.
+Permute the columns of `df` in-place, according to permutation `cols`. Elements of `cols` may be
+any valid column selection, but cannot be a combination of both.
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: What is "both"? Also, not in sync with `reordercols`.

File: docs/src/lib/functions.md [hunk 2]
```diff
@@ -62,6 +62,7 @@ show
 sort
 sort!
 unique!
+permutecols
```
Comment (nalimilan) [inline docs/src/lib/functions.md]: If functions are renamed, this needs to be updated. Same below for exports.

**Comment by nalimilan** [comment]
Having a `location` argument similar to Stata sounds like a good idea. But a keyword argument sounds more logical for that. We may also want to add a keyword argument to sort columns in alphabetic order at some point (which could be combined with `All`, `Between` or a regex).

**Comment by nalimilan** [comment]
Regarding the name, I'm not sure. "permute" sounds OK.

Also I've just realized that dplyr [uses `select`](http://www.datasciencemadesimple.com/re-arrange-re-order-column-dataframe-r-using-dplyr/) for this. In Julia the equivalent would be `select(df, cols, All())` or `select(df, Not(cols), All())`. Not sure that's really better, but probably worth considering as it simplifies the API further.

**Comment by bkamins** [comment]
@nalimilan This is an excellent observation. Actually this means that what we need to do is just to `permutecols!` to `select!`. The only thing we would loose is that `permutecols!` throws an error when an invalid permutation is passed, but I do not think it is crucial to keep it.

The reason is that to put columns in front currently you can just write:
```
select(df, All(col1, col2, All()))`
```
to put column in the back you write:
```
select(df, All(Not([col1, col2]), col1, col2))
```

the `:before` and `:after` cases would not be that simple to implement, but I think they are not super useful anyway.

The only change we could consider is to have `select(df, cols...)` be transformed to `select(df, All(cols...))`.

@nilshg and @nalimilan If you are OK with this I will put up another PR for this (i.e. deprecation of `permutecols!` and adding slurping to `select`/`select!`).

**Comment by bkamins** [comment]
See #1955. I finally decided to leave `permutecols!` as it is a more fundamental operation than `select!` (we would have to leave it internally anyway). But if you prefer we can deprecate it.

**Comment by bkamins** [comment]
I meant https://github.com/JuliaData/DataFrames.jl/pull/1964.

**Comment by nalimilan** [comment]
Why is it more fundamental?

**Comment by bkamins** [comment]
Because it has to modify internals of a `DataFrame` in-place (reordering the columns and the index). Of course I can move the logic of `permutecols!` into `select!`, but maybe it is OK to leave `permutecols!` as it additionally checks if you actually passed a permutation (which `select!` does not do and I think it should not).

**Comment by nalimilan** [comment]
So that's not really more fundamental than `select!` I'd say. Not sure whether checking that a permutation has been passed justifies having a separate function. Do you think that's really common? Julia doesn't offer a similar function for matrices, and neither dplyr not Pandas provide separate functions AFAICT.

**Comment by bkamins** [comment]
OK - I will remove it.

**Pull request closed without merging**